src/mumble/UserLocalNicknameDialog.cpp

```diff
@@ -0,0 +1,87 @@
+#include "ClientUser.h"
+#include "Database.h"
+#include "MainWindow.h"
+
+#include <QtGui/QCloseEvent>
+#include <QtWidgets/QPushButton>
+
+// We define a global macro called 'g'. This can lead to issues when included code uses 'g' as a type or parameter name
+// (like protobuf 3.7 does). As such, for now, we have to make this our last include.
+#include "Global.h"
+
+UserLocalNicknameDialog::UserLocalNicknameDialog(unsigned int sessionId,
+											 QMap< unsigned int, UserLocalNicknameDialog * > *qmUserNicknameTracker)
+	: QDialog(nullptr), m_clientSession(sessionId), m_qmUserNicknameTracker(qmUserNicknameTracker) {
+	setupUi(this);
+	qleUserLocalNickname->setAccessibleName(tr("User nickname"));
+
+	ClientUser *user = ClientUser::get(sessionId);
+	if (user) {
+		QString title = tr("Adjusting local nickname for %1").arg(user->qsName);
+		setWindowTitle(title);
+		qleUserLocalNickname->setText(user->getLocalNickname());
+		m_originalNickname = qleUserLocalNickname->text();
+	}
+
+	if (g.mw && g.mw->windowFlags() & Qt::WindowStaysOnTopHint) {
+		// If the main window is set to always be on top of other windows, we should make the
+		// nickname dialog behave the same in order for it to not get hidden behind the main window.
+		setWindowFlags(Qt::WindowStaysOnTopHint);
+	}
+}
+
+void UserLocalNicknameDialog::closeEvent(QCloseEvent *event) {
+	m_qmUserNicknameTracker->remove(m_clientSession);
+	event->accept();
+}
+
+void UserLocalNicknameDialog::present(unsigned int sessionId,
+									QMap< unsigned int, UserLocalNicknameDialog * > *qmUserNicknameTracker) {
+	if (qmUserNicknameTracker->contains(sessionId)) {
+		qmUserNicknameTracker->value(sessionId)->raise();
+	} else {
+		UserLocalNicknameDialog *userNickname = new UserLocalNicknameDialog(sessionId, qmUserNicknameTracker);
+		userNickname->show();
+		qmUserNicknameTracker->insert(sessionId, userNickname);
+	}
+}
+
+void UserLocalNicknameDialog::on_qleUserLocalNickname_textChanged(const QString &text) {
+	ClientUser *user = ClientUser::get(m_clientSession);
+	if (user) {
+		user->setLocalNickname(text);
+	}
+}
+
+void UserLocalNicknameDialog::on_qbbUserLocalNickname_clicked(QAbstractButton *button) {
+	if (button == qbbUserLocalNickname->button(QDialogButtonBox::Reset)) {
+		qleUserLocalNickname->setText("");
+	}
+	if (button == qbbUserLocalNickname->button(QDialogButtonBox::Ok)) {
+		ClientUser *user = ClientUser::get(m_clientSession);
+		if (user) {
+			if (!user->qsHash.isEmpty()) {
+				g.db->setUserLocalNickname(user->qsHash, user->getLocalNickname());
+			} else {
+				g.mw->logChangeNotPermanent(QObject::tr("Local Nickname Adjustment..."), user);
+			}
+		}
+		UserLocalNicknameDialog::close();
+	}
+	if (button == qbbUserLocalNickname->button(QDialogButtonBox::Cancel)) {
+		qleUserLocalNickname->setText(m_originalNickname);
+		UserLocalNicknameDialog::close();
+	}
+}
+
+void UserLocalNicknameDialog::reject() {
+	m_qmUserNicknameTracker->remove(m_clientSession);

```

Comment:
If I did not miss anything this here is causing a memory leak in the current implementation as the object pointed by the value stored for `m_clientSession` simply goes out of scope without having deleted the object in question (even though it was created using `new`) :thinking: 

That's why smart-pointers are preferred - with them this would be just fine as it is :)